### ISSUE TITLE
[Bug of automatic completion] Auto-completion of lists overrides parameters entered by user

### ISSUE THREAD

**Issue by LStruber** [populse/populse_mia#269]
Note: This has only be tested with controller V2

Minimum step to reproduce:
- Open your favorite project
- Add a Coregister brick (or any other brick with lists in input parameters) in the pipeline manager
- Initialize files with any files of your data browser
- **Modify** default input parameters (especially lists)
- Launch the brick and look at these parameters

The brick is launched but some of the parameters are reinitialized to their default values in the controller during automatic completion ! See image below.

**Default Coregister brick controller:**
![coregister_default](https://user-images.githubusercontent.com/86590799/164763389-c009571c-6923-415f-9d2b-5c80ef33ce51.PNG)

**Modified coregister brick controller:**
![coregister_changed](https://user-images.githubusercontent.com/86590799/164763441-c8536342-1747-438c-9ac0-c7f5c71c508a.PNG)

**Coregister brick controller when running:**
![coregister_after_completion](https://user-images.githubusercontent.com/86590799/164763592-cc70c3a3-0f17-4739-b73c-b30121df6828.PNG)

The brick is indeed launched with false (reinitialized) parameters ! This is very problematic :/

@servoz I needed to initialize some bricks by this mean in MRIQC pipeline, but as it is not working, I will hardcode these parameters in the brick, which then will become less adaptable (but working)

**Comment by servoz** [comment]
Oh damn... This is really a big bug !!!!
We need to fixe it ASAP.

**Comment by servoz** [comment]
I reproduce!

* Using Smooth from mia_processes:
  - Before starting the calculation :
![Screenshot from 2022-04-23 19-18-59](https://user-images.githubusercontent.com/37933899/164917058-de87d622-e220-41b5-a297-4ac1ba79987d.png)
  - After the calculation :
![Screenshot from 2022-04-23 19-19-28](https://user-images.githubusercontent.com/37933899/164917082-12a9a7e7-f357-4d11-9854-25b37a721e82.png)

The run was actually launched with the `fwhm == [3, 6, 6]` parameter as we can see in soma-worflow and in the history window
![Screenshot from 2022-04-23 19-20-51](https://user-images.githubusercontent.com/37933899/164917110-ef356e60-a2d4-4438-8b4e-56762d402b66.png)

When we change the value, the pipeline became blue, i had a look in the capsul pipeline_developer_view module, but i haven't found what this blue colour is (it doesn't seem to be a good sign!).

* Using Smooth from nipype :
We can observe the same issue.
This issue did not exist before. You are right, it must be due to the completion during initialisation, or else the controller. This makes me think about capsul/soma, but we should also look at populse_mia ...


During these tests, I observed another issue that is related, I think,  to the work on the history window.
There is now an entry in the History tag with the history button for nipype script file.
![Screenshot from 2022-04-23 19-35-27](https://user-images.githubusercontent.com/37933899/164933855-98600a6f-8e7b-4483-a499-e90100da5e09.png)
1/ this script should not be indexed in the project (it's an internal code file that doesn't interest the user, although this can be discussed, by the way I don't remember if we made sure not to index it like we do with the mia_processes bricks?).
2/ but more importantly I think that there shouldn't be the button to access the history for this file.
I will open another ticket about this as soon as possible.

**Comment by servoz** [comment]
Hum ... I just tested with the V1 controller... I did not observe this problem with the old controller...

So, I think the issue is strongly related to V2 controller (I think we have to investigate mainly in capsul side).

And @LStruber , maybe it is not necessary to hard code in the brick, it will be enough to use the V1 controller while waiting to fix this ticket? (I also have bricks that only work with controller V1 because of the lack of support for complex types with controllerV2).

Can you also test your bricks with the V1 controller and say if you don't see this issue with this controller too?

**Comment by LStruber** [comment]
> And @LStruber , maybe it is not necessary to hard code in the brick, it will be enough to use the V1 controller while waiting to fix this ticket? (I also have bricks that only work with controller V1 because of the lack of support for complex types with controllerV2).

I think I will implement both for the sake of simplicity (for users). One simple brick with hardcoded parameters, because the user may want to perform a "simple" registration that works in most cases (without defining parameters), and a complex registration brick with all inputs avalaible to the user.

I'll test the brick with V1 controller asap.

**Comment by servoz** [comment]
> I think I will implement both for the sake of simplicity (for users). One simple brick with hardcoded parameters, because the user may want to perform a "simple" registration that works in most cases (without defining parameters), and a complex registration brick with all inputs avalaible to the user.

Why not, but as you know you can define the default values you want for the parameters in the constructor of the class corresponding to the brick (when self.add_trait("foo", ...)). I think it's like if you hard code the value (the user can use as it is the brick) and in addition there is the possibility to change the value if the user wants. This is the idea in which the bricks in mia_processes have been coded for the moment. But as you want!

**Comment by LStruber** [comment]
You're true but I would like to have 2 default values for this brick, one for T1w file and the other for bold files. That's why I think it could be good to get a easy "T1wRegistration", a easy "BoldRegistration" and a complex "Registration", but that is of course debatable

**Comment by servoz** [comment]
> You're true but I would like to have 2 default values for this brick, one for T1w file and the other for bold files. That's why I think it could be good to get a easy "T1wRegistration", a easy "BoldRegistration" and a complex "Registration", but that is of course debatable

Ok, but ... :-))) there is also the possibility to set values when building the pipeline! Is this what you want?
For example in the spatial_preprocessing pipeline we define for the[ normalize12_2 brick, the plug write_voxel_sizes = [2.0, 2.0, 2.0]](https://github.com/populse/mia_processes/blob/master/python/mia_processes/pipelines/preprocess/spatial_preprocessing_1.py#L50).

**Comment by LStruber** [comment]
I didn't know that ! That would indeed solve the problem I guess. I'll try this way

**Comment by denisri** [comment]
(I'm coming back from vacation so I haven't looked at this ticket before)
Yes it seems to be a bug in the Controller GUI. It isn't related to completion: if you modify values in a list as you did, then open a node controller GUI on the same node, you get the initial values - or more exactly the 1st values of lists seems to be modified, not the other ones. I'll try to have a look.

**Comment by denisri** [comment]
I think I could fix the problem in soma-base. Could you try it on your side please ?

**Comment by servoz** [comment]
Oh yes, the issue is now fixed.
Thanks @denisri!
